pkg/notes/notes_v2.go

```diff
@@ -0,0 +1,249 @@
+			}
+			bar.Increment()
+			t.Done(nil)
+		}()
+
+		if t.Throttle() > 0 {
+			break
+		}
+	}
+
+	if err := t.Err(); err != nil {
+		return nil, err
+	}
+
+	bar.Finish()
+
+	return aggregator.releaseNotes, nil
+}
+
+func (g *Gatherer) buildReleaseNote(pair *commitPrPair) (*ReleaseNote, error) {
+	pr, _, err := g.client.GetPullRequest(g.context, g.options.GithubOrg, g.options.GithubRepo, pair.PrNum)
+	if err != nil {
+		return nil, err
+	}
+
+	prBody := pr.GetBody()
+
+	text, err := noteTextFromString(prBody)
+	if err != nil {
+		logrus.Debugf("sha: %s pr: %d err: %v", pair.Commit.Hash.String(), pair.PrNum, err)
+		return nil, nil
+	}
+
+	documentation := DocumentationFromString(prBody)
+
+	author := pr.GetUser().GetLogin()
+	authorURL := pr.GetUser().GetHTMLURL()
+	prURL := pr.GetHTMLURL()
+	isFeature := hasString(labelsWithPrefix(pr, "kind"), "feature")
+	noteSuffix := prettifySIGList(labelsWithPrefix(pr, "sig"))
+
+	isDuplicateSIG := false
+	if len(labelsWithPrefix(pr, "sig")) > 1 {
+		isDuplicateSIG = true
+	}
+
+	isDuplicateKind := false
+	if len(labelsWithPrefix(pr, "kind")) > 1 {
+		isDuplicateKind = true
+	}
+
+	// TODO(wilsonehusin): extract / follow original in ReleasenoteFromCommit
+	indented := strings.ReplaceAll(text, "\n", "\n  ")
+	markdown := fmt.Sprintf("%s ([#%d](%s), [@%s](%s))",
+		indented, pr.GetNumber(), prURL, author, authorURL)
+
+	if noteSuffix != "" {
+		markdown = fmt.Sprintf("%s [%s]", markdown, noteSuffix)
+	}
+
+	// Uppercase the first character of the markdown to make it look uniform
+	markdown = strings.ToUpper(string(markdown[0])) + markdown[1:]
+
+	return &ReleaseNote{
+		Commit:         pair.Commit.Hash.String(),
+		Text:           text,
+		Markdown:       markdown,
+		Documentation:  documentation,
+		Author:         author,
+		AuthorURL:      authorURL,
+		PrURL:          prURL,
+		PrNumber:       pr.GetNumber(),
+		SIGs:           labelsWithPrefix(pr, "sig"),
+		Kinds:          labelsWithPrefix(pr, "kind"),
+		Areas:          labelsWithPrefix(pr, "area"),
+		Feature:        isFeature,
+		Duplicate:      isDuplicateSIG,
+		DuplicateKind:  isDuplicateKind,
+		ActionRequired: labelExactMatch(pr, "release-note-action-required"),
+		DoNotPublish:   labelExactMatch(pr, "release-note-none"),
+	}, nil
+}
+
+func (g *Gatherer) listLeftParentCommits(opts *options.Options) ([]*commitPrPair, error) {
+	localRepository, err := git.PlainOpen(opts.RepoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// opts.StartSHA points to a tag (e.g. 1.20.0) which is on a release branch (e.g. release-1.20)
+	// this means traveling through commit history from opts.EndSHA will never reach opts.StartSHA
+
+	// the stopping point to be set should be the last shared commit between release branch and primary (master) branch
+	// usually, following the left / first parents, it would be
+	// * tag: v1.20.0, some merge commit
+	// |
+	// * Anago GCB release commit (begin branch out of release-1.20)
+	// |
+	// * last shared commit
```

Comment:
What about the changed from the master branch? Like the changes from alpha and beta releases?

Reply:
🤔 you're right, this graph is quite ambiguous. [In the original discussion with Adolfo](https://kubernetes.slack.com/archives/CN1KH4K9A/p1611884448047700?thread_ts=1611883139.043900&cid=CN1KH4K9A), my idea is roughly this:

![image](https://user-images.githubusercontent.com/14004487/106576687-4e182300-64f2-11eb-9eed-eceb8ac83743.png)

This part of the code is rewinding the pointer from `v1.20.0` to `x` (the last shared commit), and then traverse from `1.21.0-alpha.1` (which is on `master` branch) all the way to `x`.

Reply:
was that helpful to clarify?

Reply:
Ah, now I got it, thank you!